src/realm/obj.cpp

```diff
@@ -1957,18 +1959,43 @@ Obj& Obj::set_collection(ColKey col_key, CollectionType type)
     update_if_needed();
     Mixed new_val(0, type);
 
-    ArrayMixed values(_get_alloc());
+    ArrayMixed arr(_get_alloc());
     ref_type ref = to_ref(Array::get(m_mem.get_addr(), col_key.get_index().val + 1));
-    values.init_from_ref(ref);
-    auto old_val = values.get(m_row_ndx);
+    arr.init_from_ref(ref);
+    auto old_val = arr.get(m_row_ndx);
 
     if (old_val != new_val) {
-        set(col_key, Mixed(0, type));
-        // Update ref after write
-        ref = to_ref(Array::get(m_mem.get_addr(), col_key.get_index().val + 1));
-        values.init_from_ref(ref);
+        CascadeState state;
```

Comment:
The old code was not working because `values.set_key()` would cause an extra copy-on-write of the top array of the object (which is not done if REALM_ENABLE_MEMDEBUG is off). But the current Obj object would not be updated with the new copy.

Reply:
OK got it...